walkeeper/src/wal_service.rs

```diff
@@ -47,6 +39,39 @@ const LIBPQ_MSG_SIZE_OFFS: usize = 1;
 const CONTROL_FILE_NAME: &str = "safekeeper.control";
 const END_OF_STREAM: XLogRecPtr = 0;
 
+/// Read some bytes from a type that implements [`Read`] into a [`BytesMut`]
+///
+/// Will return the number of bytes read, just like `Read::read()` would.
+///
+fn read_into(r: &mut impl Read, buf: &mut BytesMut) -> io::Result<usize> {
+    // This is a workaround, because BytesMut and std::io don't play
+    // well together.
+    //
+    // I think this code needs to go away, and I'm confident that
+    // that's possible, if we are willing to refactor this code to
+    // use std::io::BufReader instead of doing buffer management
+    // ourselves.
+    //
+    // SAFETY: we already have exclusive access to self.inbuf, so
+    // there are no concurrency problems; the only risk would be
+    // accidentally exposing uninitialized parts of the buffer.
+    //
+    // We write into the buffer just past the known-initialized part,
+    // then manually increment its length by the exact number of
+    // bytes we read. So no uninitialized memory should be exposed.
+
+    let start = buf.len();
+    let end = buf.capacity();
+
+    let num_bytes = unsafe {
+        let fill_here = buf.get_unchecked_mut(start..end);
+        let num_bytes_read = r.read(fill_here)?;
+        buf.set_len(start + num_bytes_read);
+        num_bytes_read
+    };
+    Ok(num_bytes)
+}
```

Comment:
I really don't like this. I looked around for a better way, but I didn't find anything better.

I hope to do the BufReader conversion soon, so we don't need to look at this code for long.

Reply:
> I really don't like this. I looked around for a better way, but I didn't find anything better.

You could read to a temporary array, and pass it to `BytesMut::extend_from_slice`

> I hope to do the BufReader conversion soon, so we don't need to look at this code for long.

Ok.